src/services/webclient.ts

```diff
@@ -1018,28 +1036,77 @@ export class WebClientService {
 
     /**
      * Send a push message to wake up the peer.
-     * The push message will only be sent if the last push is less than 2 seconds ago.
-     */
-    private sendPush(): void {
-        // Make sure not to flood the target device with pushes
-        const minPushInterval = 2000;
-        const now = new Date();
-        if (this.lastPush !== null && (now.getTime() - this.lastPush.getTime()) < minPushInterval) {
-            this.$log.debug(this.logTag,
-                'Skipping push, last push was requested less than ' + (minPushInterval / 1000) + 's ago');
-            return;
+     *
+     * Returns the maximum expected period until the promise will be resolved,
+     * and the promise itself.
+     */
+    public sendPush(): [number, Promise<any>] {
+        // Create new session
+        if (this.pushSession === null) {
+            this.pushSession = this.pushService.createSession(this.salty.permanentKeyBytes, this.pushSessionConfig);
+
+            // Start and handle success/error
+            this.pushPromise = this.pushSession.start()
+                .then(() => this.resetPushSession(true))
+                .catch((error) => {
+                    // Reset push session
+                    this.resetPushSession(false);
+
+                    // Handle error
+                    if (error instanceof TimeoutError) {
+                        this.showDeviceUnreachableDialog();
+                    } else {
+                        this.failSession();
+                    }
+                });
+
+            // Update state
+            if (!this.$rootScope.$$phase) {
```

Comment:
Frankly, this is one of the instances where it would be much more of an antipattern to go through all possible instances to ensure it is being called in a root scope. I just didn't want to copy the huge explainer comment that we have in another place (when receiving a message). Even if it would be an expensive operation, this isn't called a lot.

> In the messenger, when clicking the retry button. This is an async function, so from my understanding this will never happen inside the digest cycle.

Just FYI: IIRC an async function is synchronous before the first `await` call. (This is different to Python's asyncio.)

> In any case, if we can't figure out anymore from where those functions can be called, we should push the hack down the stack and do this check in the state service (updateXxxState), so that the state service can be called from any task.

I'll evaluate.

Reply:
> IIRC an async function is synchronous before the first await call.

Yes, FYI, it is even synchronous if the `await` waits for an already resolved Promise. Off-topic, but just wanted to leave this note. :smile:

Reply:
@rugk Unless that is well defined in the ECMA spec, this might be implementation dependent. The above mentioned feature however is the result of syntactic sugar over Promises.